plugins/cpp/parser/src/clangastvisitor.h

```diff
@@ -921,6 +948,27 @@ class ClangASTVisitor : public clang::RecursiveASTVisitor<ClangASTVisitor>
     return true;
   }
 
+  bool VisitDeclStmt(clang::DeclStmt* ds_)
+  {
+    assert(_statements.top() == ds_ &&
+      "ds_ was expected to be the top-level statement.");
```

Comment:
I'm not sure whether this assertion holds everywhere. Unfortunately, although defining a function within another function is not possible, except for lambdae:

```cpp
int f()
{
    auto L = []() -> int {
        int X;
        return 1;
    }();
    return L;
}
```

here, `VarDecl`s `L` and `X` are both children of `DeclStmt`s, but `X` isn't top-level to the `f` function. However, it is top-level to the `operator()` of the lambda, unless having the `CompoundStmt` there is a problem...

```cpp
int f()
{
  int X = 1;
  {
    int X = 2;
    return X;
  }
}
```

Reply:
![screen](https://github.com/Ericsson/CodeCompass/assets/64849783/a32ad99d-8e39-4538-8d82-6295bff4c19f)

I have tested your code with the minor change of replacing int with a UDT that has an explicit destructor so we can actually see its usages in the Info Tree. It gets parsed perfectly without the assertion ever breaking.

Note: The 5th usage in the above screenshot at 81:5 is not currently visible in master because of #583. I used the fix in #679 locally to make it visible during testing, but that is a separate unrelated issue.
The point is that no assertions were hit.

Reply:
@whisperity Can you take a look on this?

Reply:
> @whisperity Can you take a look on this?

LGTM. 😉 It's better to have these looked at not to run into odd crashes later.